src/libraries/Microsoft.PowerFx.Core/Public/Types/TypeCheckers/AggregateTypeChecker.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Public.Types.TypeCheckers
+{
+    internal abstract class AggregateTypeChecker
+    {
+        protected ICollection<ExpressionError> _errorList;
+
+        public AggregateTypeChecker(ICollection<ExpressionError> errorList)
+        {
+            _errorList = errorList ?? throw new ArgumentNullException(nameof(errorList));
+        }
+
+        public bool Run(AggregateType sourceType, AggregateType typeToCheck)
+        {
+            if (sourceType == null || typeToCheck == null)
+            {
+                return false;
+            }
+            else if (sourceType._type.Kind != typeToCheck._type.Kind)
+            {
+                _errorList.Add(new ExpressionError()
+                {
+                    Kind = ErrorKind.Validation,
+                    Severity = ErrorSeverity.Critical,
+                    Message = $"Type mismatch between source and target types. Expected: {sourceType._type.GetKindString()}, found {typeToCheck._type.GetKindString()}."
+                });
+                return false;
+            }
+
+            return IsMatch(sourceType, typeToCheck);
+        }
+
+        public abstract bool IsMatch(AggregateType sourceType, AggregateType typeToCheck);
+
+        public abstract bool IsFieldMatch(FormulaType sourceField, FormulaType fieldToCheck);
+    }
+
+    internal class StrictAggregateTypeChecker : AggregateTypeChecker
+    {
+        public StrictAggregateTypeChecker(ICollection<ExpressionError> errorList)
+            : base(errorList)
+        {
+        }
+
+        public override bool IsMatch(AggregateType sourceType, AggregateType typeToCheck)
+        {
+            var sourceFieldCount = sourceType.FieldNames.Count();
+            var typeToCheckFieldCount = typeToCheck.FieldNames.Count();
+            if (sourceFieldCount != typeToCheckFieldCount)
+            {
+                if (sourceFieldCount < typeToCheckFieldCount)
+                {
+                    var extraFields = string.Join(", ", typeToCheck.FieldNames.Except(sourceType.FieldNames));
+                    _errorList.Add(new ExpressionError()
+                    {
+                        Kind = ErrorKind.Validation,
+                        Severity = ErrorSeverity.Critical,
+                        Message = $"Type mismatch between source and target record types. Given type has extra fields: {extraFields}."
```

Comment:
>$"Type mismatch between source [](http://example.com/codeflow?start=34&length=30)

The error detail here is great. We will need to add to resx so they can be localized. 

Does existing Dtype comparison have similar messages we can reuse? (or maybe as a followup, it should start leveraging this!)

Reply:
Added to resx file